### PR TITLE
refactor: move reference resolver into shared package

### PR DESCRIPTION
**Pull request by justinsb**

### CODE REVIEW

**Review by yuwenma — COMMENTED** [review]
/lgtm
/approve
/hold

feel like we can either merge these two files to a `visitreferences.go` or a sub package `reference` for easy discoverable purpose. What do you think? Feel free to unhold.

**Comment by justinsb** [comment]
The reason I kept them separate is because I think we'll want the "walk a proto" functionality in other places also.  e.g. implementing update_mask

**Comment by justinsb** [comment]
/assign @yuwenma 

Can we move this forwards?  Would be good to standardize reference resolving

File: pkg/controller/direct/common/refs.go
Comment (yuwenma) [inline pkg/controller/direct/common/refs.go:1]: Do you want to move the MonitoringAlertPolicy specific functions out of this file?

**Review by yuwenma — COMMENTED** [review]
/approve
/lgtm

/hold 

One suggestion. Not a blocker. Thank you for this!

File: pkg/controller/direct/common/refs.go
Comment (justinsb) [reply pkg/controller/direct/common/refs.go:1]: I think not yet ... I read a tweet the other day: https://x.com/KentBeck/status/250733358307500032?lang=en

So this is making the change easy, but not yet making the easy change.  The easy changes are to unify all the resolvers, including picking up the Reference interface :-)

**Comment by justinsb** [comment]
Ah, but I've lost LGTM in a rebase anyway :-)

**Comment by yuwenma** [comment]
/lgtm


> Ah, but I've lost LGTM in a rebase anyway :-)

No, you got it in seconds.

**Pull request merged**